rust-src/concordium_base/src/id/sigma_protocols/vcom_eq.rs

```diff
@@ -0,0 +1,413 @@
+        ro.append_message("g_bar", &self.g_bar)
+    }
+
+    fn get_challenge(&self, challenge: &Challenge) -> Self::ProtocolChallenge {
+        C::scalar_from_bytes(challenge)
+    }
+
+    fn commit_point<R: rand::Rng>(
+        &self,
+        csprng: &mut R,
+    ) -> Option<(Self::CommitMessage, Self::ProverState)> {
+        let n = self.gis.len();
+        let rtilde = C::generate_non_zero_scalar(csprng);
+        let mut alphas = Vec::with_capacity(n);
+        let mut rtildes = BTreeMap::new();
+        let mut ais = Vec::with_capacity(self.comms.len());
+        // let mut i = 0;
+        for i_usize in 0..n {
+            let i = i_usize.try_into().ok()?;
+            let alpha_i = C::generate_non_zero_scalar(csprng);
+            alphas.push(alpha_i);
+            if self.comms.contains_key(&i) {
+                let rtilde_i = C::generate_non_zero_scalar(csprng);
+                rtildes.insert(i, rtilde_i);
+                let ai = multiexp(&[self.g_bar, self.h_bar], &[alpha_i, rtilde_i]);
+                ais.push(ai);
+            }
+            // i += 1;
+        }
+        // for the factor h^{r_tilde}
+        alphas.push(rtilde);
+        let mut gis = self.gis.clone();
+        gis.push(self.h);
+        let a = multiexp(&gis, &alphas); // h^rtilde \prod g_i^(alpha_i)
+        alphas.pop(); // remove rtilde from alphas again.
+        Some(((a, ais), (alphas, rtilde, rtildes)))
+    }
+
+    fn generate_witness(
+        &self,
+        secret: Self::SecretData,
+        state: Self::ProverState,
+        challenge: &Self::ProtocolChallenge,
+    ) -> Option<Self::ProverWitness> {
+        let (alphas, rtilde, rtildes) = state;
+        let (xis, r, ris) = secret;
+        let n = xis.len();
+        if alphas.len() != n || rtildes.len() != ris.len() || ris.len() != self.comms.len() {
+            return None;
+        }
+        let mut sis = Vec::with_capacity(n);
+        let mut tis = BTreeMap::new();
+        for (i_usize, (ref alpha_i, ref xi)) in izip!(alphas, xis).enumerate() {
+            let mut si = *challenge;
+            si.mul_assign(xi);
+            si.negate();
+            si.add_assign(alpha_i);
+            sis.push(si);
+            let i = i_usize.try_into().ok()?;
+            if let (Some(rtilde_i), Some(ri)) = (rtildes.get(&i), ris.get(&i)) {
+                let mut ti = *challenge;
+                ti.mul_assign(ri);
+                ti.negate();
+                ti.add_assign(rtilde_i);
+                tis.insert(i, ti);
+            }
+        }
+        let mut t = *challenge;
+        t.mul_assign(&r);
+        t.negate();
+        t.add_assign(&rtilde);
+        Some(Witness { sis, t, tis })
+    }
+
+    fn extract_point(
+        &self,
+        challenge: &Self::ProtocolChallenge,
+        witness: &Self::ProverWitness,
+    ) -> Option<Self::CommitMessage> {
+        let Witness { sis, t, tis } = witness;
+        let maybe_fit: Result<IndexType, _> = (sis.len() - 1).try_into();
```

Comment:
Do we have a problem here if `sis` is empty? Can that happen in verification?

Reply:
Added an extra check. `sis` should never be empty.